crates/networking/manager/src/gossipsub/validate/beacon_block.rs

```diff
@@ -0,0 +1,291 @@
+        };
+
+        let Some(parent_state) = store
+            .db
+            .beacon_state_provider()
+            .get(block.message.parent_root)?
+        else {
+            return Err(anyhow!("failed to get parent state"));
+        };
+
+        (parent_block, parent_state)
+    };
+
+    // Validate parent block [block.message.parent_root]
+    match validate_beacon_block(beacon_chain, cached_db, &parent_block, &parent_state, true).await?
+    {
+        ValidationResult::Accept => {}
+        ValidationResult::Ignore(reason) => {
+            return Ok(ValidationResult::Ignore(reason));
+        }
+        ValidationResult::Reject(reason) => {
+            return Ok(ValidationResult::Reject(reason));
+        }
+    };
+
+    let validator = if let Some(validator) = latest_state_in_db
+        .validators
+        .get(block.message.proposer_index as usize)
+    {
+        validator
+    } else {
+        return Ok(ValidationResult::Reject("Validator not found".to_string()));
+    };
+
+    cached_db.seen_proposer_signature.write().await.put(
+        AddressSlotIdentifier {
+            address: validator.public_key.clone(),
+            slot: block.message.slot,
+        },
+        block.signature.clone(),
+    );
+
+    for signed_bls_execution_change in block.message.body.bls_to_execution_changes.iter() {
+        let validator = &latest_state_in_db.validators
+            [signed_bls_execution_change.message.validator_index as usize];
+
+        cached_db.seen_bls_to_execution_signature.write().await.put(
+            AddressSlotIdentifier {
+                address: validator.public_key.clone(),
+                slot: block.message.slot,
+            },
+            signed_bls_execution_change.message.clone(),
+        );
+    }
+
+    Ok(ValidationResult::Accept)
+}
+
+pub async fn validate_beacon_block(
+    beacon_chain: &BeaconChain,
+    cached_db: &CachedDB,
+    block: &SignedBeaconBlock,
+    state: &BeaconState,
+    is_parent: bool,
+) -> anyhow::Result<ValidationResult> {
+    let store = beacon_chain.store.lock().await;
+
+    // [IGNORE] The block is not from a future slot.
+    if block.message.slot > store.get_current_slot()? {
+        return Ok(ValidationResult::Ignore(
+            "Block is from a future slot".to_string(),
+        ));
+    }
+
+    // [IGNORE] The block is from a slot greater than the latest finalized slot.
+    if block.message.slot
+        <= compute_start_slot_at_epoch(store.db.finalized_checkpoint_provider().get()?.epoch)
+    {
+        return Ok(ValidationResult::Ignore(
+            "Block is from a slot greater than the latest finalized slot".to_string(),
+        ));
+    }
+
+    let validator =
+        if let Some(validator) = state.validators.get(block.message.proposer_index as usize) {
+            validator
+        } else {
+            return Ok(ValidationResult::Reject("Validator not found".to_string()));
+        };
```

Comment:
```suggestion
    let Some(validator) = state.validators.get(block.message.proposer_index as usize) else {
            return Ok(ValidationResult::Reject("Validator not found".to_string()));
        };
```

^ we can simplify the code to this